[quant][graph] Add quantized batch_norm2d_relu to graph mode

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36622 [quant][graph] Graph mode quantization support for sigmoid
* **#36552 [quant][graph] Add quantized batch_norm2d_relu to graph mode**
* #36692 [quant][graph] Add quantized batch_norm2d support to graph mode
* #36691 [quant][graph] Add useQuantizable function

Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21075253](https://our.internmc.facebook.com/intern/diff/D21075253)

supriyar committed:
[quant][graph] Add quantized batch_norm2d_relu to graph mode
Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1405,6 +1405,42 @@ def forward(self, x):
+                self.bn = torch.nn.BatchNorm2d(3).to('cpu', torch.float)
+
+            def forward(self, x):
+                return F.relu(self.bn(x))
```
can you add a `inplace` argument here as well?

supriyar replied:
Added a comment explaining that it is not supported.

jerryzh168 replied:
you can still test for the case the `inplace` argument is a constant I think

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1405,6 +1405,50 @@ def forward(self, x):
+                self.bn = torch.nn.BatchNorm2d(3).to('cpu', torch.float)
+
+            def forward(self, x):
+                return F.relu(self.bn(x), True)
```
can you make inplace an argument to the module?

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1405,6 +1405,50 @@ def forward(self, x):
+        class FuncRelu(torch.nn.Module):
+            def __init__(self):
+                super(FuncRelu, self).__init__()
+                self.bn = torch.nn.BatchNorm2d(3).to('cpu', torch.float)
```
`cpu` is default

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1405,6 +1405,50 @@ def forward(self, x):
+            def forward(self, x):
+                return F.relu(self.bn(x), False)
+
+        data = [(torch.rand((1, 3, 10, 10), dtype=torch.float), torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        model = self._test_op_impl(FuncRelu, data, "quantized::batch_norm2d_relu")
+        FileCheck().check_not("aten::batch_norm") \
+                   .check_not("aten::relu") \
+                   .run(model.graph)
+        model = self._test_op_impl(FuncInplaceRelu, data, "quantized::batch_norm2d_relu")
+        for inplace in [True, False]:
+            m = torch.jit.script(M(inplace).eval())
+            m = prepare_script(m, {'': default_qconfig}, True)
+            data = torch.rand((1, 3, 10, 10), dtype=torch.float)
+            m(data)
+            m = convert_script(m, False)
+            FileCheck().check("quantized::batch_norm2d_relu") \
+                       .check_not("aten::batch_norm") \
+                       .check_not("aten::relu") \
+                       .check_not("aten::relu_") \
+                       .run(m.graph)
+
```
can you remove code duplication here:?

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1405,6 +1405,42 @@ def forward(self, x):
+                   .run(model.graph)
+        for inplace in [True, False]:
+            m = torch.jit.script(M(inplace).eval())
+            m = prepare_script(m, {'': default_qconfig}, True)
+            data = torch.rand((1, 3, 10, 10), dtype=torch.float)
+            m(data)
+            m = convert_script(m, False)
```
use `_test_op_impl`?

jerryzh168 replied:
can you merge all the checks in one for loop?

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1163,9 +1163,9 @@ class TestQuantizeScriptPTSQOps(JitTestCase):
     for individual ops end to end.
     """
-    def _test_op_impl(self, SingleOpModule, data, quantized_op):
+    def _test_op_impl(self, SingleOpModule, data, quantized_op, **kwargs):
```
can you change this to take a module instance? then you can initialize the module before passing into this function.

jerryzh168 reviewed (changes requested):
please refactor `_test_op_impl`

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1400,11 +1400,47 @@ def forward(self, x):
+                return F.relu(self.bn(x), True)
+
+        data = [(torch.rand((1, 3, 10, 10), dtype=torch.float), torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        for Model in [BNRelu, BNFuncRelu, BNFuncInplaceRelu]:
+            for inplace in [True, False]:
+                model = self._test_op_impl(Model(inplace), data, "quantized::batch_norm2d_relu")
```
please change this loop as well, we don't need to loop through [True, False] anymore

supriyar replied:
We still need to test BNRelu model for different inplace argument.

supriyar replied:
Good catch! Updated it

supriyar committed:
Update on "[quant][graph] Add quantized batch_norm2d_relu to graph mode"


Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good, thanks!
